core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -885,13 +867,37 @@ class LocalCacheStatus {
    LocalCacheStatus(CacheJoinInfo joinInfo,
                     CacheTopologyHandler handler,
                     PartitionHandlingManager phm,
-                    int stableMembersSize) {
+                    int stableMembersSize,
+                    ComponentRegistry cr) {
       this.joinInfo = joinInfo;
       this.handler = handler;
       this.partitionHandlingManager = phm;
-      this.stable = stableMembersSize > 0 ? new CompletableFuture<>() : CompletableFutures.completedFalse();
       this.knownMembers = Collections.emptyList();
       this.stableMembersSize = stableMembersSize;
+      this.stable = stableMembersSize > 0 ? new CompletableFuture<>() : CompletableFutures.completedFalse();
+
+      // If the cache needs recovery, we need to handle the storage after the stable topology is installed.
+      // In cases where it was not safely recovered, we need to clear the stores.
+      if (needRecovery()) {
```

Comment:
Debating if this if here is necessary or not. Would it be possible to have a private store with data without a persistent state? Should we let it run just to be safe?